packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -650,6 +709,47 @@ export class RemotedFluidDataStoreContext extends FluidDataStoreContext {
 
         return this.details;
     }
+
+    /**
+     * This should be called when running garbage collection to get details needed for
+     * idetifying which fluid objects to collect.
+     */
+    protected async getGarbageCollectioDetails(): Promise<IGarbageCollectionDetails> {
+        if (!this.garbageCollectionDetails) {
+            let tree: ISnapshotTree | null;
+            let garbageCollectionDetails: IGarbageCollectionDetails | undefined;
+
+            if (typeof this.initSnapshotValue === "string") {
+                const commit = (await this.storage.getVersions(this.initSnapshotValue, 1))[0];
+                tree = await this.storage.getSnapshotTree(commit);
+            } else {
+                tree = await this.initSnapshotValue;
+            }
+
+            const localReadAndParse = async <T>(id: string) => readAndParse<T>(this.storage, id);
+            if (tree) {
+                const loadedSummary = await this.summarizerNode.loadBaseSummary(tree, localReadAndParse);
+                tree = loadedSummary.baseSummary;
+            }
+
+            // Read the garbage collection blob if its available.
+            if (tree !== null && tree.blobs[garbageCollectionBlobKey] !== undefined) {
+                garbageCollectionDetails =
+                    await localReadAndParse<IGarbageCollectionDetails>(tree.blobs[garbageCollectionBlobKey]);
+            }
+
+            /**
+             * If there is no snapshot tree or no garbage collection attributes blob, set isRootDataStore to true.
+             * This will ensure that data stores in older documents are not garbage collected incorrectly. This may
+             * lead to additional roots in the document but they won't break.
+             */
+            this.garbageCollectionDetails = {
+                isRootDataStore: garbageCollectionDetails?.isRootDataStore ?? true,
```

Comment:
Assuming you will keep adding to garbageCollectionDetails, I'd not add defaults here, but rather initialize garbageCollectionDetails  with defaults where you declare it - at the top of the function. That way you will not need to test every member one by one